File: jmlee/week7/searchinsertposition/searchinsertposition.go
package searchinsertposition

func searchInsert(nums []int, target int) int {
	return find(nums, target, 0, len(nums)-1)
}

func find(nums []int, target int, left int, right int) int {
	center := ((right - left) / 2) + left
	if nums[center] == target {
		return center
	} else if right-left <= 1 {
		if nums[right] == target {
			return right
		} else if target < nums[left] {
			return left
		} else if target > nums[right] {
			return right + 1
		} else {
			return right
		}
	} else if nums[center] < target {
		return find(nums, target, center, right)
	} else if nums[center] > target {
		return find(nums, target, left, center)
	} else {
		return -1
	}
}

File: carrotfarm/src/p2018_08_28/SearchInsertPosition.java
package p2018_08_28;

import static org.junit.Assert.assertEquals;

import org.junit.Test;

public class SearchInsertPosition {

	public int searchInsert(int[] nums, int target) {

		int l = 0;
		int r = nums.length;

		int mid = 0;
		while(l < r) {

			mid = (l+r) / 2;
			if (nums[mid] == target) {
				return mid;
			}else if (nums[mid] > target) {
				r = mid;
			}else {
				l = mid;
			}
			if (r==0) {
				return 0;
			}
			if (l==nums.length-1) {
				return nums.length;
			}
			if (nums[l] < target && r-l == 1) {
				return r;
			}
		}
		return 0;
	}

	@Test
	public void test() {

//		assertEquals(searchInsert(new int[] {1,3,5,6}, 5), 2);
//		assertEquals(searchInsert(new int[] {1,3,5,6}, 2), 1);
//		assertEquals(searchInsert(new int[] {1,3,5,6}, 7), 4);
//		assertEquals(searchInsert(new int[] {1,3,5,6}, 0), 0);
		assertEquals(searchInsert(new int[] {1,3}, 1), 0);
	}
}
